### PR TITLE
SIH: Publish Aux Global Positon

### PR DESCRIPTION
**Pull request by bresch**
### Solved Problem
I need to test Aux Global Position (AGP) in sim.

### Solution
Implement SensorAgpSim to publish AGP based on groundtruth data.

image

### CODE REVIEW

File: src/modules/simulation/sensor_agp_sim/parameters.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  */
+PARAM_DEFINE_INT32(SENS_EN_AGPSIM, 0);
+
+/**
+ * AGP failure mode
+ *
+ * @group Simulator
+ * @min 0
+ * @max 3
+ * @bit 0 Constant
```
[the remark this answers is not in the document]
Comment (bresch) [reply src/modules/simulation/sensor_agp_sim/parameters.c:51]: Yes, I wasn't sure if "frozen", "stuck" or "constant" would be more appropriate; I can change it to "stuck" if you find it clearer

File: src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+{
+	// generate white Gaussian noise sample with std=1
+	// from BlockRandGauss.hpp
+	static float V1, V2, S;
+	static bool phase = true;
+	float X;
+
+	if (phase) {
+		do {
+			float U1 = (float)rand() / (float)RAND_MAX;
```
[the remark this answers is not in the document]
Comment (bresch) [reply src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp:69]: I dumbly copy-pasted the `SensorGpsSim` class. 
I think it requires some refactoring as it's nonsense to re-declare that function in every class. I don't understand if there is was a real reason to not have a common lib.

File: src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+		} else {
+			_position_bias.zero();
+		}
+
+		double latitude = _measured_lla.latitude_deg() + math::degrees((double)generate_wgn() * 2.0 /
+				  CONSTANTS_RADIUS_OF_EARTH);
+		double longitude = _measured_lla.longitude_deg() + math::degrees((double)generate_wgn() * 2.0 /
+				   CONSTANTS_RADIUS_OF_EARTH);
+		double altitude = (double)(_measured_lla.altitude() + (generate_wgn() * 0.5f));
```
[the remark this answers is not in the document]
Comment (bresch) [reply src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp:130]: Yes, maybe as parameters? So we can then change the noise level during the simulation.

File: src/modules/simulation/sensor_agp_sim/parameters.c [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * @min 0
+ * @max 1
+ * @group Sensors
+ * @value 0 Disabled
+ * @value 1 Enabled
+  */
+PARAM_DEFINE_INT32(SENS_EN_AGPSIM, 0);
+
+/**
+ * AGP failure mode
```
[the remark this answers is not in the document]
Comment (bresch) [reply src/modules/simulation/sensor_agp_sim/parameters.c:46]: Hmm, I think the original objective for this simulation module is to be re-used in multiple simulators, not only SIH.

[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/simulation/sensor_agp_sim/parameters.c:46]: Ah yes ofc, it's also used in GZ. Disregard comemnt.

File: src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp]: Not a bad idea, but also okay to leave it hard-coded but properly exposed as constant for now. Up to you, you have a better feeling for how we use it atm.

File: src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfuhrer) [reply src/modules/simulation/sensor_agp_sim/SensorAgpSim.cpp:69]: okok, all good for now.

**Comment by hamishwillee** [comment]
This adds new parameters [SIM_AGP_FAIL](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SIM_AGP_FAIL) and [SENS_EN_AGPSIM](https://docs.px4.io/main/en/advanced_config/parameter_reference.html#SENS_EN_AGPSIM) and 

So I see AGP = Aux Global Position - so this provides you with a second source of global position so you can test estimators work OK? 

Does this need further docs - say as a sub heading of https://docs.px4.io/main/en/simulation/#sitl-simulation-environment ?

I'd say if you want anyone else other than you to use this then probably a mention. But if it is very much internal, then maybe not.

@bresch ?

**Pull request merged**